lib/codecs/src/decoding/format/avro.rs

```diff
@@ -0,0 +1,302 @@
+            .unwrap();
+        AvroDeserializer {
+            schema,
+            strip_schema_id_prefix: self.avro_options.strip_schema_id_prefix,
+        }
+    }
+
+    /// The data type of events that are accepted by `AvroDeserializer`.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        // TODO: convert avro schema to vector schema.
+        match log_namespace {
+            LogNamespace::Legacy => {
+                let mut definition = schema::Definition::empty_legacy_namespace()
+                    .unknown_fields(vrl::value::Kind::any());
+
+                if let Some(timestamp_key) = log_schema().timestamp_key() {
+                    definition = definition.try_with_field(
+                        timestamp_key,
+                        vrl::value::Kind::any().or_timestamp(),
+                        Some("timestamp"),
+                    );
+                }
+                definition
+            }
+            LogNamespace::Vector => schema::Definition::new_with_default_metadata(
+                vrl::value::Kind::any(),
+                [log_namespace],
+            ),
+        }
+    }
+}
+
+impl From<&AvroDeserializerOptions> for AvroSerializerOptions {
+    fn from(value: &AvroDeserializerOptions) -> Self {
+        Self {
+            schema: value.schema.clone(),
+        }
+    }
+}
+/// Apache Avro serializer options.
+#[configurable_component]
+#[derive(Clone, Debug)]
+pub struct AvroDeserializerOptions {
+    /// The Avro schema.
+    #[configurable(metadata(
+        docs::examples = r#"{ "type": "record", "name": "log", "fields": [{ "name": "message", "type": "string" }] }"#
+    ))]
+    pub schema: String,
+
+    /// for avro datum encoded in kafka messages, the bytes are prefixed with the schema id.  Set this to true to strip the schema id prefix.
+    /// According to [Confluent Kafka's document](https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#wire-format)
+    pub strip_schema_id_prefix: bool,
+}
+
+/// Serializer that converts bytes to an `Event` using the Apache Avro format.
+#[derive(Debug, Clone)]
+pub struct AvroDeserializer {
+    schema: apache_avro::Schema,
+    strip_schema_id_prefix: bool,
+}
+
+impl AvroDeserializer {
+    /// Creates a new `AvroDeserializer`.
+    pub const fn new(schema: apache_avro::Schema, strip_schema_id_prefix: bool) -> Self {
+        Self {
+            schema,
+            strip_schema_id_prefix,
+        }
+    }
+
+    /// Deserializes the given bytes, which will always produce a single `LogEvent`.
+    pub fn parse_single(
+        &self,
+        bytes: Bytes,
+        _log_namespace: LogNamespace,
+    ) -> vector_common::Result<LogEvent> {
+        let bytes = if self.strip_schema_id_prefix {
```

Comment:
💬  Do we want to accept empty `bytes` here? If yes, let's return `Ok` immediately. If not, return an error.

Reply:
AFAIK, [`null`](https://avro.apache.org/docs/1.11.1/specification/#primitive-types) is a AVRO type, and avro decoder schema will check it which I think is better than return `Ok(())`.

This is a demo which bad branch will fail.

```rust
use apache_avro::Schema;
use serde::{Deserialize, Serialize};


#[derive(Debug, Serialize,Deserialize,Clone)]
struct Test {
    a: i64,
    b: String,
}

#[derive(Debug, Serialize,Deserialize,Clone)]
struct Empty {
}

#[derive(Debug, Serialize,Deserialize,Clone)]
struct Test2 {
    a:i32,
    b:(),
}

fn main() {
    let raw_schema = r#"
    {
        "type": "null"
    }
    "#;
    let schema = Schema::parse_str(raw_schema).unwrap();

    let test = ();
    
    // Serialize data
    let value = apache_avro::to_value(test).unwrap();
    let value = value.resolve(&schema).unwrap();
    let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
    println!("serialize. data: {:?}, len: {}", bytes, bytes.len());

    // Deseriallize data
    let r = bytes.clone();
    let mut r = r.as_slice();
    let value = apache_avro::from_avro_datum(&schema, &mut r,None).unwrap();
    let value:() = apache_avro::from_value::<()>(&value).unwrap();
    println!("deseriallize. data: {:?}", value);

    // Bad deseriallize
    let raw_schema = r#"
    {
        "type": "int"
    }
    "#;
    let r = bytes.clone();
    let mut r = r.as_slice();
    let schema = Schema::parse_str(raw_schema).unwrap();
    let value = apache_avro::from_avro_datum(&schema, &mut r,None).unwrap();
    let value:() = apache_avro::from_value::<()>(&value).unwrap();
    println!("bad deseriallize. data:{:?}", value);
}
```

Reply:
This makes sense, thank you! 

Let's add an early check:
```
if bytes.is_empty() {
    return Ok(smallvec![]);
}
```